### PR TITLE
test: Extend salvage test for wallet tool

### PR DESCRIPTION
**Pull request by maflcko**
Extend the test to check the balance remains the same before and after the salvage

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Test seems to fail:
https://bitcoinbuilds.org/index.php?job=b27a270b-c15a-4a48-bf46-26b3cc36a464

**Comment by jonatack** [comment]
Concept ACK.

Seeing this error:
```
2020-06-05T18:01:34.518000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_hf5xwzyh
2020-06-05T18:01:34.975000Z TestFramework (INFO): Testing that various invalid commands raise with specific error messages
2020-06-05T18:01:35.809000Z TestFramework (INFO): Calling wallet tool info, testing output
2020-06-05T18:01:36.535000Z TestFramework (INFO): Generating transaction to mutate wallet
2020-06-05T18:01:39.036000Z TestFramework (INFO): Calling wallet tool info after generating a transaction, testing output
2020-06-05T18:01:39.238000Z TestFramework (INFO): Calling wallet tool create on an existing wallet, testing output
2020-06-05T18:01:41.947000Z TestFramework (INFO): Starting node with arg -wallet=foo
2020-06-05T18:01:42.716000Z TestFramework (INFO): Calling getwalletinfo on a different wallet ("foo"), testing output
2020-06-05T18:01:42.975000Z TestFramework (INFO): Check salvage
2020-06-05T18:01:51.543000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 119, in main
    self.run_test()
  File "test/functional/tool_wallet.py", line 245, in run_test
    self.test_salvage()
  File "test/functional/tool_wallet.py", line 232, in test_salvage
    self.assert_tool_output('', '-wallet=salvage', 'salvage')
  File "test/functional/tool_wallet.py", line 52, in assert_tool_output
    assert_equal(stderr, '')
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 46, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(bitcoin-wallet: wallet/walletdb.cpp:289: auto ReadKeyValue(CWallet *, CDataStream &, CDataStream &, CWalletScanState &, std::string &, std::string &)::(anonymous class)::operator()(CWalletTx &, bool) const: Assertion `new_tx' failed.
 == )
2020-06-05T18:01:51.595000Z TestFramework (INFO): Stopping nodes
```

File: test/functional/tool_wallet.py
Comment (achow101) [inline test/functional/tool_wallet.py:234]: I had to add `self.nodes[0].rescanblockchain(0)` to make this pass after merging with #19457

**Comment by achow101** [comment]
#19805 is a potential solution to the test failure as it just avoids deserializing those records entirely.

#19793 is a another potential solution to the test failure where the assert is removed.

**Comment by ryanofsky** [comment]
> #19805 is a potential solution to the test failure as it just avoids deserializing those records entirely.
> 
> #19793 is a another potential solution to the test failure where the assert is removed.

Maybe I should look at the test to understand better, but both of these changes are avoiding failures when there are duplicate transactions in the database. But we still don't know how duplicate transactions are winding up in the database to begin with?

**Comment by achow101** [comment]
> But we still don't know how duplicate transactions are winding up in the database to begin with?

The duplicate transactions are showing up in the records the salvage pulls from the database. This doesn't mean the database actually contains duplicate transactions but is actually more of an artifact of the aggressive salvage.

**Comment by MarcoFalke** [comment]
Rebased

File: test/functional/tool_wallet.py
Comment (ryanofsky) [inline test/functional/tool_wallet.py:232]: In commit "test: Add salvage test for wallet tool" (fa6154b3aa904eb39e6e537ff99106b6e022701b)


It seems the test is failing on this line due to stderr output https://travis-ci.org/github/bitcoin/bitcoin/jobs/735969416#L2582:

```
Salvage: WARNING: Number of keys in data does not match number of values.
Salvage: WARNING: Unexpected end of file while reading salvage output.
WARNING: WalletBatch::Recover skipping key: CDataStream::read(): end of data: unspecified iostream_category error
WARNING: WalletBatch::Recover skipping key: CDataStream::read(): end of data: unspecified iostream_category error
WARNING: WalletBatch::Recover skipping key: Error reading wallet database: CPubKey/CPrivKey corrupt
```

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa6154b3aa904eb39e6e537ff99106b6e022701b

Test looks right to me but there is an travis failure on one platform

File: test/functional/tool_wallet.py
Comment (MarcoFalke) [reply test/functional/tool_wallet.py:232]: #20151

**Review by ryanofsky — APPROVED** [review]
Code review ACK 66660e440d888e0cd1a16a769b7c06fc0d8e8345. Only change since last review is rebase. Also travis seems to pass now.

I'm confused about why the test was failing before https://github.com/bitcoin/bitcoin/pull/19078#discussion_r509798005 but succeeding now. The bug report #20151 just points back to this PR for steps to reproduce (which I guess no longer work?) and doesn't mention whether the problem happens reliably or is maybe random or platform specific.

**Comment by MarcoFalke** [comment]
Please don't merge this. The failure still happens intermittently.

**Comment by meshcollider** [comment]
@MarcoFalke how often does the intermittent test failure occur? I've rebased this on [a branch](https://github.com/meshcollider/bitcoin/commits/202109_salvage) which incorporates a few other salvage things, and can't hit a failure after multiple re-runs locally. I don't know if that's because the failure is rare or fixed.

**Comment by MarcoFalke** [comment]
How often did you try? I am running 

```
while test/functional/tool_wallet.py --legacy-wallet ; do ( echo 1 >> /tmp/runs )  ; done
```

and it fails after 36 tries:

```
$ wc -l  /tmp/runs 
36 /tmp/runs

**Comment by MarcoFalke** [comment]
On my next try I got more than 600 passes and no fail :thinking:

**Comment by achow101** [comment]
Are you still working on this?

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

**Comment by MarcoFalke** [comment]
Ok, I'll rebase

**Pull request closed without merging**
